compiler/rustc_lint/src/unused.rs

```diff
@@ -1033,7 +1033,6 @@ impl EarlyLintPass for UnusedParens {
         if let ast::TyKind::Paren(r) = &ty.kind {
             match &r.kind {
                 ast::TyKind::TraitObject(..) => {}
-                ast::TyKind::BareFn(b) if b.generic_params.len() > 0 => {}
                 ast::TyKind::ImplTrait(_, bounds) if bounds.len() > 1 => {}
                 ast::TyKind::Array(_, len) => {
                     self.check_unused_delims_expr(
```

Comment:
I just realized, that's a bug :grin: it causes us to not lint on 
```rust
fn main() {
    let _x: ([u32; 3]);
    //      ^^^^^^^^^^ should lint
    let _y: [u8; (3)];
    //           ^^^ should lint
    let _z: ([u8; (3)]);
    //            ^^^ does lint
}
```
may make sense as part of this PR. HAve to move the `ast::TyKind::Array` branch to the same level as the `if let ast::TyKind::Paren(r) = &ty.kind {`, not inside of `TyKind::Paren`.

If not I am going to open a separate issue for this.

Reply:
Seems we find some lint error in `std` 😂:

```diff
error: unnecessary parentheses around const expression
   --> library/std/src/io/error/repr_bitpacked.rs:377:27
    |
377 | static_assert!(@usize_eq: (TAG_MASK & TAG_SIMPLE_MESSAGE), TAG_SIMPLE_MESSAGE);
    |                           ^                             ^
    |
    = note: `-D unused-parens` implied by `-D warnings`
help: remove these parentheses
    |
377 - static_assert!(@usize_eq: (TAG_MASK & TAG_SIMPLE_MESSAGE), TAG_SIMPLE_MESSAGE);
377 + static_assert!(@usize_eq: TAG_MASK & TAG_SIMPLE_MESSAGE, TAG_SIMPLE_MESSAGE);
    |

error: unnecessary parentheses around const expression
   --> library/std/src/io/error/repr_bitpacked.rs:378:27
    |
378 | static_assert!(@usize_eq: (TAG_MASK & TAG_CUSTOM), TAG_CUSTOM);
    |                           ^                     ^
    |
help: remove these parentheses
    |
378 - static_assert!(@usize_eq: (TAG_MASK & TAG_CUSTOM), TAG_CUSTOM);
378 + static_assert!(@usize_eq: TAG_MASK & TAG_CUSTOM, TAG_CUSTOM);
    |

```